subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java

```diff
@@ -40,13 +47,48 @@ public AbstractDaemonCompiler(File daemonWorkingDir, Compiler<T> delegate, Compi
     @Override
     public WorkResult execute(T spec) {
         DaemonForkOptions daemonForkOptions = toDaemonOptions(spec);
-        CompilerDaemon daemon = compilerDaemonFactory.getDaemon(daemonWorkingDir, daemonForkOptions);
-        CompileResult result = daemon.execute(delegate, spec);
+        WorkerDaemon daemon = compilerDaemonFactory.getDaemon(CompilerDaemonServer.class, daemonWorkingDir, daemonForkOptions);
+        WorkerDaemonResult result = daemon.execute(adapter(delegate), spec);
         if (result.isSuccess()) {
             return result;
         }
         throw UncheckedException.throwAsUncheckedException(result.getException());
     }
 
+    private CompilerWorkerAdapter<T> adapter(Compiler<T> compiler) {
+        return new CompilerWorkerAdapter<T>(compiler);
+    }
+
     protected abstract DaemonForkOptions toDaemonOptions(T spec);
+
+    private static class CompilerWorkerAdapter<T extends CompileSpec> implements WorkerDaemonAction<T> {
+        private final Compiler<T> compiler;
+
+        CompilerWorkerAdapter(Compiler<T> compiler) {
+            this.compiler = compiler;
+        }
+
+        @Override
+        public WorkerDaemonResult execute(T spec) {
+            return new WorkerDaemonResult(compiler.execute(spec).getDidWork(), null);
+        }
+
+        @Override
+        public String getDescription() {
+            return compiler.getClass().getName();
+        }
+    }
+
+    // TODO Come up with a better way to set up the worker implementation classpath
+    // This is a hack to get the appropriate classpath on the worker implementation classpath for compiler daemons.
+    // The classpath is derived from the implementation class and when this is WorkerDaemonServer, we get the classpath
+    // from the classloader in the Gradle Core API classloader scope (which only contains certain jars).  Using this
+    // class causes the classpath to be inferred from the Gradle API scope classloader instead so that we get the necessary
+    // jars for a compiler daemon.
+    public static class CompilerDaemonServer extends WorkerDaemonServer {
```

Comment:
No, the `WorkerDaemonExecutor.classpath()` method adds things to the _application_ classpath.  This is about getting the Gradle compiler classes (`CompilerWorkerAdapter`, `CompileSpec`, etc) onto the _implementation_ classpath.  We can't put them on the application classpath because they will drag along and pollute it with other stuff like groovy jars, scala jars, etc - we need to be able to set the application classpath to specific versions of things like groovy.  The implementation classpath is derived from the server-side implementation class, which for the public api is `WorkerDaemonServer` but that won't infer the correct classpath to get all of the compiler classes we need.  That's what this class is for - a server-side implementation class that can be used to infer the correct implementation classpath specifically for compiler daemons.  This isn't great, but the `WorkerProcess` stuff currently tries to hide the implementation classpath from us, so this is a simple solution until we can provide a better way to inject the implementation classpath.